Model quirk For Bosch BTH-RM230Z room thermostat

cygnusb opened the pull request:
This fixes handling of set_temperature, if device is used in newer Z2M, which requires to set target_temp_high and target_temp_low instead of temperature.
Backward compatiblity is ensured by checking model, platform (mqtt), and if heat/cool hvac_modes are present.
Fixes KartoffelToby/better_thermostat#1401

## Motivation:

Z2M has changed handling of hvac modes (https://github.com/Koenkk/zigbee-herdsman-converters/commit/2101914e85d864246694a1f8d334a92781ffe9bc) resulting mode heat and cool now present. This requires now to **set target_temp_low** and **target_temp_high** instead of **temperature** in action call. 

## Changes:

Add model quirk BTH-RM230Z.py

## Related issue (check one):

- [x] fixes #1401 
- [ ] there is no related issue ticket

## Checklist (check one):

- [ ] I did not change any code (e.g. documentation changes)
- [x] The code change is tested and works locally.

## Test-Hardware list (for code changes)

HA Version: 2024.10.3
Zigbee2MQTT Version: 1.40.2
TRV Hardware: Bosch Room thermostat II 230V RBSH-RTH0-ZB-EU and model BTH-RM230Z

## New device mappings

- [ ] I avoided any changes to other device mappings
- [x] There are no changes in `climate.py`

KartoffelToby commented:
@cygnusb sry for the long wait, shoud be in the next Release, thanks for your work :)

hablahab commented:
Thanks @cygnusb and @KartoffelToby, unfortunately this fixes only the mains powered device BTH-RM230Z not the battery powered one BTH-RM. What makes it even worse, the battery powered one does not identify correctly as model "BTH-RM" but instead as "Battery model" only.
I fixed it by adding the model fix Battery_model.py with the xact same content as BTH-RM230Z.py only switching line 25 to:

`    if model == "Battery model":`

Since the damn thing has a whitespace in the model, I also had to modify model_quirks.py adding after line 11:

`    model = model.replace(" ", "_")`

I hope no other Bosch devices identify simply as "Battery model"

hablahab commented:
Quick Update: Just noticed the model_fix works only one way. Changing the Temperature for the Better Thermostat sets the Temperature on the Wall Thermostat but not the other way around which I would consider much more important.

cygnusb commented:
> Thanks @cygnusb and @KartoffelToby, unfortunately this fixes only the mains powered device BTH-RM230Z not the battery powered one BTH-RM. What makes it even worse, the battery powered one does not identify correctly as model 

Can you send me the MQTT Discovery data of your battery model? I only have mains powered ones.

Device - MQTT Info

Required Info should be below the climate instance. Something like

`homeassistant/climate/0x<ID>/climate/config`

should contain in the Payload something like:

```
device:
  identifiers:
    - zigbee2mqtt_0x<ID>
  manufacturer: Bosch
  model: Room thermostat II 230V (BTH-RM230Z)
  via_device: zigbee2mqtt_bridge_0x<ID>
  name: Kueche Thermostat
  ...
```

cygnusb commented:
> Quick Update: Just noticed the model_fix works only one way. Changing the Temperature for the Better Thermostat sets the Temperature on the Wall Thermostat but not the other way around which I would consider much more important.

Yes it seems that setting the temperature does not get communicated to better_thermostat at all. Possibly because the temperature is using another attribute. 

Since all my devices are in child lock and only controlled through Home Assistant UI I did not check this. Might take a look at this. But this might need more than a model fix.

cygnusb commented:
Ok, thanks. It seems that Z2M reports this as modename (typecode) and modelname already contains (). 

The issue is in https://github.com/KartoffelToby/better_thermostat/blob/358cbda73d737051fba56b012bb0e289c9913efe/custom_components/better_thermostat/utils/helpers.py#L459

which only captures the text within the first () resulting in the internal Better Thermostat model "Battery model" and not in BTH-RM. Solution is to fix this in utils/helpers.py only taking text in between the last () and create another model quirk BTH-RM for the Battery model.

Let me also check if I can fix the local control on the thermostat to report back to BT.

cygnusb commented:
I have added multiple pull requests, which (in combination) should fix all issues. @hablahab you can find all changes merged in [cygnusb:bosch_all_fixes](https://github.com/cygnusb/better_thermostat/tree/bosch_all_fixes). This should fix the detection of the Battery Model, add a model quirk for the battery model. This also allows controlling the temperature locally on the device. Changes are communicated back to BT in HA and grouped devices are also updated.

KartoffelToby commented:
@cygnusb thanks for your work,

I'll merge your PRs into master and release a 1.7.0 Beta with this and lots of other changes soon
